src/Compilers/CSharp/Portable/Binder/PatternExplainer.cs

```diff
@@ -0,0 +1,512 @@
+            {
+                if (evaluations.IsEmpty &&
+                    constraints.All(t => t switch
+                    {
+                        (BoundDagValueTest _, _) => true,
+                        (BoundDagRelationalTest _, _) => true,
+                        (BoundDagExplicitNullTest _, false) => true,
+                        (BoundDagNonNullTest _, true) => true,
+                        _ => false
+                    }) &&
+                    ValueSetFactory.ForType(input.Type) is { } fac)
+                {
+                    // All we have are numeric constraints. Process them to compute a value not covered.
+                    var remainingValues = fac.AllValues;
+                    foreach (var constraint in constraints)
+                    {
+                        var (test, sense) = constraint;
+                        switch (test)
+                        {
+                            case BoundDagValueTest v:
+                                addRelation(BinaryOperatorKind.Equal, v.Value);
+                                break;
+                            case BoundDagRelationalTest r:
+                                addRelation(r.Relation, r.Value);
+                                break;
+                        }
+                        void addRelation(BinaryOperatorKind relation, ConstantValue value)
+                        {
+                            var filtered = fac.Related(relation, value);
+                            if (!sense)
+                                filtered = filtered.Complement();
+                            remainingValues = remainingValues.Intersect(filtered);
+                        }
+                    }
+
+                    if (remainingValues.Complement().IsEmpty)
+                        return "_";
+
+                    return SampleValueString(remainingValues, input.Type, requireExactType: requireExactType);
+                }
+
+                return null;
+            }
+
+            // Handle the special case of a recursive pattern
+            string tryHandleRecursivePattern()
+            {
+                if (constraints.IsEmpty && evaluations.IsEmpty)
+                    return null;
+
+                string typeName = requireExactType ? input.Type.ToDisplayString() : null;
+
+                foreach (var con in constraints)
+                {
+                    switch (con)
+                    {
+                        case (test: BoundDagNonNullTest _, sense: true):
+                        case (test: BoundDagExplicitNullTest _, sense: false):
+                            // not-null tests are implicitly incorporated into a recursive pattern
+                            break;
+                        default:
+                            return null;
+                    }
+                }
+
+                string deconstruction = null;
+                var properties = new Dictionary<Symbol, string>();
```

Comment:
>properties [](start = 20, length = 10)

It seems like the properties `Dictionary<,>` could be replaced by a `StringBuilder` that builds up the property string.

Reply:
I'm OK with this formulation.

---
In reply to: [443080115](https://github.com/dotnet/roslyn/pull/44702#discussion_r443080115) [](ancestors = 443080115)